File: 0386-lexicographical-numbers/0386-lexicographical-numbers.py
class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        def dfs(start, end):
            for i in range(start, min(n + 1, end + 1)):
                ans.append(i)
                dfs(i * 10, i * 10 + 9)
          
          
        ans = []
        dfs(1, 9)
        return ans

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.js
/**
 * @param {number} n
 * @return {number[]}
 */
var lexicalOrder = function(n) {
  const ans = [];
  
  const pushFromTo = (start, end) => {
    while (start <= end && start <= n) {
      ans.push(start);
      pushFromTo(start * 10, start * 10 + 9);
      start++;
    }
  }
  
  pushFromTo(1, 9);
  
  return ans;
};